File: DSA/src/Questions/String/ChunkWords.java
package Questions.String;

import java.util.ArrayList;

class ChunkWords {

    public static ArrayList<String> solve(String str) {
        if (str == null) {
            return new ArrayList<>();
        }
        int space = 0;
        int start = 0;
        int n = str.length();
        ArrayList<String> result = new ArrayList<>();
        int limit = 160;
        if (n <= limit) {
            result.add(str);
            return result;
        }
        // O(N)
        for (int i = 0; i < str.length(); i++) {
            char ch = str.charAt(i);

            if (ch == ' ') {
                space = i;
            }

            int buffer = 6;
            if (i - start >= limit - buffer) {

                // Reset the Starts
                if (i < n - 1 && str.charAt(i) != ' ') {
                    // Take upto the space
                    result.add(str.substring(start, space));
                    start = space + 1;
                } else {
                    // Take upto i
                    result.add(str.substring(start, i));
                    start = i + 1;
                }
            }
        }

        if (result.size() == 1) {
            return result;
        }
        // O(n)
        for (int i = 0; i < result.size(); i++) {
            String index = " (" + i + "/" + result.size() + ")";
            // System.out.println(index);
            result.set(i, result.get(i) + index);
        }

        return result;
    }

    public static void main(String[] args) {
        String str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.";
        ArrayList<String> result = solve(str);
        System.out.println(result);
        // Null
        result = solve(null);
        System.out.println(result);
        str = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed nec orci sem. Proin nunc ex, suscipit sagittis tempor ac, tempor ut ante. Phasellus in dapibus nam.";
        result = solve(str);
        System.out.println(result);
        str = "Lorem ipsum dolor sit amet";
        result = solve(str);
        System.out.println(result);
    }

}

File: DSA/Python/Questions/String/ChunkWords.py
"""
# Question: Chunk Words
# Link: (Custom Problem)

# Problem Statement:
# Given a text string, split it into chunks where each chunk has a maximum size limit.
# The chunks should be split at word boundaries to maintain readability.
# Each chunk should be labeled with its position in the sequence (e.g., "1/3").

# Example:
# Input: text = "Lorem ipsum dolor sit amet, consectetur adipiscing elit"
# Output: ["Lorem ipsum (1/2)", "dolor sit amet (2/2)"]

# Key Requirements:
# - Maximum chunk size: 160 characters
# - Split at word boundaries only
# - Add position indicators for multiple chunks
"""

from typing import List


class TextChunker:
    def chunk_text(self, text: str, limit: int = 35) -> List[str]:
        """
        Split text into chunks with size constraints
        Time Complexity: O(n) where n is the length of text
        Space Complexity: O(n) for storing chunks
        """
        if not text:
            return []

        chunks = []
        start = 0
        last_space = 0
        n = len(text)

        # Process text character by character
        for i in range(n):
            # Track the last space position
            if text[i] == " ":
                last_space = i

            # Check if current segment exceeds limit
            if i - start >= limit - 6:  # 6 characters reserved for chunk indicator
                if text[i] != " ":
                    # Split at last space
                    chunks.append(text[start:last_space])
                    start = last_space + 1
                else:
                    # Split at current position
                    chunks.append(text[start:i])
                    start = i + 1

        # Add remaining text as final chunk
        if start < n:
            chunks.append(text[start:])

        # Add chunk indicators if multiple chunks exist
        if len(chunks) > 1:
            total = len(chunks)
            chunks = [f"{chunk} ({i+1}/{total})" for i, chunk in enumerate(chunks)]

        return chunks


def main():
    test_cases = [
        {
            "text": "Lorem ipsum dolor sit amet",
            "expected": ["Lorem ipsum dolor sit amet"],
        },
        {
            "text": "Lorem ipsum dolor sit amet " * 3,
            "expected": [
                "Lorem ipsum dolor sit amet (1/3)",
                "Lorem ipsum dolor sit amet (2/3)",
                "Lorem ipsum dolor sit amet  (3/3)",
            ],
        },
        {"text": "", "expected": []},
    ]

    chunker = TextChunker()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Input Text: {test['text'][:50]}...")
        result = chunker.chunk_text(test["text"])
        print(f"Result: {result}")
        print(f"Expected: {test['expected']}")
        assert result == test["expected"], f"Test case {i + 1} failed"


if __name__ == "__main__":
    main()
